core/src/wallet/bitcoin/transaction_builders/BitcoinLikeUtxoPicker.cpp

```diff
@@ -158,94 +157,113 @@ namespace ledger {
-                return unit;
-            });
+            }
         }
 
         BitcoinLikeGetUtxoFunction
         BitcoinLikeUtxoPicker::createFilteredUtxoFunction(const BitcoinLikeTransactionBuildRequest &request,
+                                                          const std::shared_ptr<BitcoinLikeKeychain> &keychain,
                                                           const BitcoinLikeGetUtxoFunction &getUtxo) {
-            auto minUtxoAmount = getCurrency().bitcoinLikeNetworkParameters.value().DustAmount;
-            return [=] () -> Future<std::vector<std::shared_ptr<api::BitcoinLikeOutput>>> {
-                return getUtxo().map<std::vector<std::shared_ptr<api::BitcoinLikeOutput>>>(getContext(), [=] (const std::vector<std::shared_ptr<api::BitcoinLikeOutput>>& utxo) {
-                    std::vector<std::shared_ptr<api::BitcoinLikeOutput>> filtered;
-                    auto isExcluded = [&] (const std::shared_ptr<api::BitcoinLikeOutput>& output) -> bool {
-                        if (output->getValue()->toLong() < minUtxoAmount) {
-                            return true;
-                        }
-                        for (auto& o : request.excludedUtxo) {
-                            auto hash = std::get<0>(o);
-                            auto index = std::get<1>(o);
-                            if (output->getTransactionHash() == hash && output->getOutputIndex() == index)
-                                return true;
-                        }
-                        return false;
+            auto minAmount = getCurrency().bitcoinLikeNetworkParameters.value().DustAmount;
+            return [=] () -> Future<std::vector<BitcoinLikeBlockchainExplorerUtxo>> {
+                return getUtxo().map<std::vector<BitcoinLikeBlockchainExplorerUtxo>>(getContext(), [=] (auto const &utxos) {
+                    auto const isExcluded = [&] (auto const &currentUtxo) {
+                        auto const compareUtxo = [&currentUtxo](auto const &utxo) {
+                            return currentUtxo.transactionHash == utxo.transactionHash &&
+                                currentUtxo.index == utxo.outputIndex;
+                        };
+
+                        // NOTE: This logic can be move to the SQL request itself but since we iterate through
+                        // the entire vector to compare the UTXO amount, this is fine to apply the filter(s) here
+                        // The filter are sorted by ascending time order
+                        //
+                        auto const &excludedUtxos = request.excludedUtxos;
+
+                        return currentUtxo.address.isEmpty()
+                            || currentUtxo.value.toLong() < minAmount
+                            || !keychain->contains(currentUtxo.address.getValue())
+                            || std::find_if(excludedUtxos.cbegin(), excludedUtxos.cend(), compareUtxo) != excludedUtxos.cend();
```

Comment:
[FAR] Current filtering is done in O(|utxos| * |excludedUtxos|). If we define `excludedUtxos` as `unordered_set<pair<string, int32_t>>` we can make find in line 259 for a constant time.